examples/vision/knowledgedistillation.py

```diff
@@ -0,0 +1,224 @@
+        self.alpha = alpha
+        self.T = T
+
+    def train_step(self, data):
+        # Unpack data
+        x, y = data
+
+        # Forward pass of teacher
+        t_pred = self.teacher(x, training=False)
+
+        with tf.GradientTape() as tape:
+            # Forward pass of student
+            s_pred = self.student(x, training=True)
+
+            # Compute losses
+            s_loss = self.s_loss_fn(y, s_pred)
+            d_loss = self.d_loss_fn(
+                tf.nn.softmax(t_pred / self.T, axis=1),
+                tf.nn.softmax(s_pred / self.T, axis=1),
+            )
+            loss = self.alpha * s_loss + (1 - self.alpha) * d_loss
+
+        # Compute gradients
+        trainable_vars = self.student.trainable_variables
+        gradients = tape.gradient(loss, trainable_vars)
+
+        # Update weights
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        # Update the metrics configured in `compile()`.
+        self.compiled_metrics.update_state(y, s_pred)
+
+        # Return a dict of performance
+        return {
+            **{"s_loss": s_loss, "d_loss": d_loss},
+            **{m.name: m.result() for m in self.metrics},
+        }
+
+    def test_step(self, data):
+        # Unpack the data
+        x, y = data
+
+        # Compute predictions
+        y_pred = self.student(x, training=False)
+
+        # Calculate the loss
+        s_loss = self.s_loss_fn(y, y_pred)
+
+        # Update the metrics.
+        self.compiled_metrics.update_state(y, y_pred)
+
+        # Return a dict of performance
+        return {**{"s_loss": s_loss}, **{m.name: m.result() for m in self.metrics}}
+
+
+"""
+## Create student and teacher models
+Initialy, we create a teacher model and a smaller student model. Both models are
+convolutional neural networks and created using `Sequential()`, but could be any model
+created with the Keras API.
+"""
+
+# Create the teacher
+teacher = keras.Sequential(
+    [
+        keras.Input(shape=(28, 28, 1)),
+        layers.Conv2D(256, (3, 3), strides=(2, 2), padding="same"),
+        layers.LeakyReLU(alpha=0.2),
+        layers.MaxPooling2D(pool_size=(2, 2), strides=(1, 1), padding="same"),
+        layers.Conv2D(512, (3, 3), strides=(2, 2), padding="same"),
+        layers.Flatten(),
+        layers.Dense(10),
+    ],
+    name="teacher",
+)
+
+# Create the student
+student = keras.Sequential(
+    [
+        keras.Input(shape=(28, 28, 1)),
+        layers.Conv2D(16, (3, 3), strides=(2, 2), padding="same"),
+        layers.LeakyReLU(alpha=0.2),
+        layers.MaxPooling2D(pool_size=(2, 2), strides=(1, 1), padding="same"),
+        layers.Conv2D(32, (3, 3), strides=(2, 2), padding="same"),
+        layers.Flatten(),
+        layers.Dense(10),
+    ],
+    name="student",
+)
+
+"""
+## Prepare dataset
+The dataset used for training the teacher and distilling the teacher is
+[MNIST](https://keras.io/api/datasets/mnist/). Both the student and teacher are trained
```

Comment:
Could we use CIFAR10, and accordingly larger convnets?

Reply:
Yes, this example should be compatible with most classical image classification tasks (given the right dataset and models), and minor adjustments could adapt it to a variety of other model types.

I have added a note about this in the example.